### PR TITLE
Optimise push in S3 driver

### PR DESCRIPTION
**Pull request by milosgajdos**
This PR attempts to optimise the performance of image push in the S3 driver. There are 2 main changes:
* we refactor the S3 driver `writer` where instead of using simple bytes slices for `ready` and `pending` parts which get constantly appended data into them causing unnecessary allocations we use optimised bytes buffers; we make sure these are used efficiently
* we introduce a memory pool that is used for allocating the byte buffers introduced above

These changes should alleviate high memory pressure on the push path to S3 especially when large images (layers) are pushed to the S3 store.

Related PR:
* https://github.com/distribution/distribution/pull/3798

### CODE REVIEW

**Comment by milosgajdos** [comment]
@corhere would you mind having a look at this WIP? You were on a review for https://github.com/distribution/distribution/pull/3798 and this PR builds, but more importantly it improves the buffer of the underlying S3 `writer`.

File: registry/storage/driver/s3-aws/s3.go [hunk 1]
```diff
@@ -1380,61 +1427,84 @@ func (w *writer) Write(p []byte) (int, error) {
...
-				if err != nil {
-					w.size += int64(n)
-					return n, err
-				}
-			} else {
-				w.pendingPart = append(w.pendingPart, p...)
-				n += len(p)
-				p = nil
+		// we've read all bytes and stil havent filled the ready buffer
+		if err == io.EOF {
```
Comment (corhere) [inline registry/storage/driver/s3-aws/s3.go:1449]:
```suggestion
		if reader.Len() == 0 {
```

**Review by corhere — COMMENTED** [review]
I feel like I'm missing something. What's the point of double-buffering if the flush operation is blocking? Is a small final chunk really that big of a deal that it's worth doubling memory usage to avoid it?

Why can't the write loop be as simple as this? Copy to buffer until full, flush buffer, repeat until there is nothing left to copy.
```go
var n int
defer func() { w.size += int64(n) }()
reader := bytes.NewReader(p)
for reader.Len() > 0 {
    m, _ := w.buf.ReadFrom(reader)
    n += m
    if w.buf.Available() == 0 {
        if err := w.flushPart(); err != nil {
            return n, err
        }
    }
}
return n, nil
```

File: registry/storage/driver/s3-aws/s3.go [hunk 2]
```diff
@@ -1251,43 +1261,73 @@ func (d *driver) getStorageClass() *string {
...
-	key         string
-	uploadID    string
-	parts       []*s3.Part
-	size        int64
-	readyPart   []byte
-	pendingPart []byte
-	closed      bool
-	committed   bool
-	cancelled   bool
+	ctx       context.Context
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1299]: Yes, I am aware of this indeed, and yes it's sucky; the reason why the `context` is the `writer` field is that the `writer` makes a bunch of S3 API calls that require passing the context to them (unless we opt not to use the `*WithContext()` SDK methods at the expense of not being able to propagate the cancellations from the driver calls).

File: registry/storage/driver/s3-aws/s3.go [hunk 3]
```diff
@@ -1251,43 +1261,73 @@ func (d *driver) getStorageClass() *string {
...
+type completedParts []*s3.CompletedPart
+
+func (a completedParts) Len() int           { return len(a) }
+func (a completedParts) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a completedParts) Less(i, j int) bool { return *a[i].PartNumber < *a[j].PartNumber }
+
+// buffer is a static size bytes buffer.
+type buffer struct {
+	data []byte
+	pos  int
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1273]: Ah yeah, this is an excellent suggestion. I'm not sure why I didnt think of this. Thanks.

File: registry/storage/driver/s3-aws/s3.go [hunk 4]
```diff
@@ -1380,61 +1427,84 @@ func (w *writer) Write(p []byte) (int, error) {
...
+		w.pool.Put(w.pending)
+	}()
+
 	return w.flushPart()
 }
 
-func (w *writer) Cancel(ctx context.Context) error {
+// NOTE(milosgajdos): writer maintains its own context
+// passed to its constructor from storage driver.
+func (w *writer) Cancel(_ context.Context) error {
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1500]: Yes, I commented above. Wish we could somehow link related comments.

File: registry/storage/driver/s3-aws/s3.go [hunk 5]
```diff
@@ -1450,33 +1520,36 @@ func (w *writer) Commit() error {
...
 	// to the completedUploadedParts slice used to complete the Multipart upload.
 	if len(w.parts) == 0 {
-		resp, err := w.driver.S3.UploadPart(&s3.UploadPartInput{
+		resp, err := w.driver.S3.UploadPartWithContext(w.ctx, &s3.UploadPartInput{
 			Bucket:     aws.String(w.driver.Bucket),
 			Key:        aws.String(w.key),
 			PartNumber: aws.Int64(1),
 			UploadId:   aws.String(w.uploadID),
-			Body:       bytes.NewReader(nil),
+			Body:       bytes.NewReader([]byte{}),
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1552]: This is just me being silly for some reason I can't think of.

File: registry/storage/driver/s3-aws/s3.go [hunk 6]
```diff
@@ -1512,34 +1587,39 @@ func (w *writer) Commit() error {
...
-		w.readyPart = append(w.readyPart, w.pendingPart...)
-		w.pendingPart = nil
+
+	buf := bytes.NewBuffer(w.ready.data[:w.ready.pos])
+	if w.pending.pos < int(w.driver.ChunkSize) {
+		if _, err := buf.Write(w.pending.data[:w.pending.pos]); err != nil {
+			return err
+		}
+		w.pending.pos = 0
 	}
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1600]: You are right, I have accidentally removed `w.driver.MultipartCombineSmallPart` and apparently never realized it - Thanks. We need better test coverage for this.

File: registry/storage/driver/s3-aws/s3.go [hunk 7]
```diff
@@ -1251,43 +1261,73 @@ func (d *driver) getStorageClass() *string {
...
+	key       string
+	uploadID  string
+	parts     []*s3.Part
+	size      int64
+	ready     *buffer
+	pending   *buffer
+	closed    bool
+	committed bool
+	cancelled bool
+	pool      *sync.Pool
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1310]: The reason is, when we `Close` the `writer`, we return both writer buffers back to the driver pool: https://github.com/distribution/distribution/pull/4066/files#diff-42ab115e8f2c28fa837e174c081ffe3542e13f5fbd58271f06b179c1123c7905R1488-R1493

File: registry/storage/driver/s3-aws/s3.go [hunk 8]
```diff
@@ -1380,61 +1427,84 @@ func (w *writer) Write(p []byte) (int, error) {
...
 	if w.closed {
 		return fmt.Errorf("already closed")
 	}
 	w.closed = true
+
+	defer func() {
+		w.ready.pos = 0
+		w.pool.Put(w.ready)
+		w.pending.pos = 0
+		w.pool.Put(w.pending)
```
Comment (corhere) [inline registry/storage/driver/s3-aws/s3.go:1492]:
```suggestion
		w.ready.pos = 0
		w.driver.pool.Put(w.ready)
		w.pending.pos = 0
		w.driver.pool.Put(w.pending)
```

File: registry/storage/driver/s3-aws/s3.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1310]: addressed in https://github.com/distribution/distribution/pull/4066/commits/62f80a633a019ffd7d9a940953272b29c9f87a85

**Comment by milosgajdos** [comment]
I've updated the `writer` implementation on all the awesome feedback @corhere  Ngl, felt mildly embarrassed to misread some of your comments 🤦‍♂️ Thanks for the commit suggestions that nudge me to re-read them 😄 

> I feel like I'm missing something. What's the point of double-buffering if the flush operation is blocking? Is a small final chunk really that big of a deal that it's worth doubling memory usage to avoid it?

This is a billion-dollar question that we may want to address in the future 😄 I have done a bit of `git` archaeology attempting to find some reasoning in the PR comments/git commits but couldn't find anything.

There is one more comment I need to think about a bit: lazy allocation of the `pending` buffer; maybe we can address that one in the follow-up, too.

File: registry/storage/driver/s3-aws/s3.go [hunk 9]
```diff
@@ -1380,61 +1443,83 @@ func (w *writer) Write(p []byte) (int, error) {
...
+
+		// try filling up the pending parts buffer
+		offset, err = w.pending.ReadFrom(reader)
+		if err != nil {
+			return 0, err
+		}
+		n += int(offset)
+
+		// we didn't fill up pending buffer, flush
+		if w.pending.Len() <= w.pending.Cap() {
```
Comment (corhere) [inline registry/storage/driver/s3-aws/s3.go:1476]: `len(a) <= cap(a)` is an invariant, therefore this condition is always true. It also does not match the conditions under which the old code would flush the buffer. The logic of the old code was "if there is enough remaining data in `p` to fill the pending buffer to capacity, fill the pending buffer and flush."
```suggestion
		// we filled up pending buffer, flush
		if w.pending.Len() == w.pending.Cap() {
```

File: registry/storage/driver/s3-aws/s3.go [hunk 10]
```diff
@@ -1300,7 +1357,7 @@ func (w *writer) Write(p []byte) (int, error) {
 	// If the last written part is smaller than minChunkSize, we need to make a
 	// new multipart upload :sadface:
 	if len(w.parts) > 0 && int(*w.parts[len(w.parts)-1].Size) < minChunkSize {
-		var completedUploadedParts completedParts
+		completedUploadedParts := make(completedParts, 0, len(w.parts))
 		for _, part := range w.parts {
 			completedUploadedParts = append(completedUploadedParts, &s3.CompletedPart{
```
Comment (corhere) [inline registry/storage/driver/s3-aws/s3.go:1362]:
```suggestion
		completedUploadedParts := make(completedParts, len(w.parts))
		for i, part := range w.parts {
			completedUploadedParts[i] = &s3.CompletedPart{
```
for consistency with lines 1546-8

File: registry/storage/driver/s3-aws/s3.go [hunk 11]
```diff
@@ -1499,47 +1587,54 @@ func (w *writer) Commit() error {
...
 		PartNumber: partNumber,
-		Size:       aws.Int64(int64(len(w.readyPart))),
+		Size:       aws.Int64(int64(w.ready.Len())),
 	})
-	w.readyPart = w.pendingPart
-	w.pendingPart = nil
+
+	// swap buffers and reset pending buffer
+	w.ready, w.pending = w.pending, w.ready
+	w.pending.Clear()
```
Comment (corhere) [inline registry/storage/driver/s3-aws/s3.go:1637]: Nit: I find it confusing to follow along with the code when the buffer referenced through `w.ready` throughout this function is cleared through `w.pending`. Do the buffer swap last to cut down on the cognitive complexity.
```suggestion
	// reset the flushed buffer and swap buffers
	w.ready.Clear()
	w.ready, w.pending = w.pending, w.ready
```

File: registry/storage/driver/s3-aws/s3.go [hunk 12]
```diff
@@ -1499,47 +1587,54 @@ func (w *writer) Commit() error {
...
 	})
 	if err != nil {
 		return err
 	}
+
 	w.parts = append(w.parts, &s3.Part{
 		ETag:       resp.ETag,
 		PartNumber: partNumber,
-		Size:       aws.Int64(int64(len(w.readyPart))),
+		Size:       aws.Int64(int64(w.ready.Len())),
```
Comment (corhere) [inline registry/storage/driver/s3-aws/s3.go:1632]: This is incorrect: the size has to be the length of `buf` before `UploadPart` to account for the increased part size when the `MultipartCombineSmallPart` optimization is in use.

**Review by corhere — CHANGES_REQUESTED** [review]
Another mystery: why is the storage driver not simply implemented in terms of https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.45.16/service/s3/s3manager#Uploader?

File: registry/storage/driver/s3-aws/s3.go [hunk 13]
```diff
@@ -1343,22 +1402,26 @@ func (w *writer) Write(p []byte) (int, error) {
...
+			// reset uploaded parts
 			w.parts = nil
-			w.readyPart, err = io.ReadAll(resp.Body)
-			if err != nil {
+			w.ready.Clear()
+
+			_, err = io.ReadFull(resp.Body, w.ready.data)
+			if err != nil && err != io.ErrUnexpectedEOF {
 				return 0, err
 			}
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1421]: Oh, wow, this is a great catch! GH suggestion diff truncates the code where I `Clear` the buffer.

File: registry/storage/driver/s3-aws/s3.go (hunk 9 above)
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1479]: Yah. Great catch.

File: registry/storage/driver/s3-aws/s3.go (hunk 12 above)
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go]: https://github.com/distribution/distribution/pull/4066/commits/ef4b24e879cedca6acd0c51c22ae27e7dbae11a0

File: registry/storage/driver/s3-aws/s3.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1318]: I dont mind, happy to rip it out; as for the AWS SDK calls that accept this context as a param, do you prefer dropping the `*WithContext()` calls or temporarily passing `context.TODO()` to them with some comment?

[the remark this answers is not in the document]
Comment (corhere) [reply registry/storage/driver/s3-aws/s3.go:1318]: I'd prefer reverting the calls so they don't show up as changed lines in the PR diff. It's easier to review that way.

File: registry/storage/driver/s3-aws/s3.go [hunk 14]
```diff
@@ -1251,43 +1261,90 @@ func (d *driver) getStorageClass() *string {
 	return aws.String(d.StorageClass)
 }
 
+type completedParts []*s3.CompletedPart
+
+func (a completedParts) Len() int           { return len(a) }
+func (a completedParts) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a completedParts) Less(i, j int) bool { return *a[i].PartNumber < *a[j].PartNumber }
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1268]: I like keeping type implementations grouped together instead of interleaving them - happy to change

File: registry/storage/driver/s3-aws/s3.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1318]: Let me do that then.

[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go]: Done in https://github.com/distribution/distribution/pull/4066/commits/0463ade9e8edabc3cf595ea735c11b705e6ba919

Curious about how you wanna go about this. I'm worried you'll suggest some wild context wrappers 🙃

[the remark this answers is not in the document]
Comment (corhere) [reply registry/storage/driver/s3-aws/s3.go]: > I'm worried you'll suggest some wild context wrappers 🙃

[Don't worry, I won't.](https://github.com/golang/go/wiki/CodeReviewComments#contexts:~:text=Don%27t%20create%20custom%20Context%20types%20or%20use%20interfaces%20other%20than%20Context%20in%20function%20signatures.)

**Review by corhere — COMMENTED** [review]
LGTM! I'll approve once you've squashed the PR. Don't worry about retaining my authorship credit on the suggestions.

I left a couple of optional suggestions for further improving the code. Emphasis on optional.

File: registry/storage/driver/s3-aws/s3.go [hunk 15]
```diff
@@ -1380,51 +1444,70 @@ func (w *writer) Write(p []byte) (int, error) {
...
+		if w.pending.Len() == w.pending.Cap() {
+			if err := w.flush(); err != nil {
+				return n, err
 			}
 		}
+
+		// we've read all bytes
+		if reader.Len() == 0 {
+			break
+		}
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1488]: Nice spot. Will update!

File: registry/storage/driver/s3-aws/s3.go [hunk 16]
```diff
@@ -1499,47 +1575,56 @@ func (w *writer) Commit() error {
...
 	})
 	if err != nil {
-		w.driver.S3.AbortMultipartUpload(&s3.AbortMultipartUploadInput{
+		if _, aErr := w.driver.S3.AbortMultipartUpload(&s3.AbortMultipartUploadInput{
 			Bucket:   aws.String(w.driver.Bucket),
 			Key:      aws.String(w.key),
 			UploadId: aws.String(w.uploadID),
-		})
+		}); aErr != nil {
+			return fmt.Errorf("failed to complete upload: %v: abort error: %v", err, aErr)
```
Comment (corhere) [inline registry/storage/driver/s3-aws/s3.go:1583]:
```suggestion
			return errors.Join(err, aErr)
```

**Comment by milosgajdos** [comment]
Alright, squashed, PTAL @corhere @thaJeztah @Jamstah

**Review by corhere — APPROVED** [review]
LGTM! I see a couple of minor formatting nits, but I am going to pretend I didn't notice them.

File: registry/storage/driver/s3-aws/s3.go [hunk 17]
```diff
@@ -1251,43 +1261,73 @@ func (d *driver) getStorageClass() *string {
...
 	}
 	return &writer{
+		ctx:      ctx,
 		driver:   d,
 		key:      key,
 		uploadID: uploadID,
 		parts:    parts,
 		size:     size,
+		ready:    d.NewBuffer(),
+		pending:  d.NewBuffer(),
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply registry/storage/driver/s3-aws/s3.go:1342]: ☝️ I see this was a "non-blocking" comment, but curious if you gave this some consideration.

File: registry/storage/driver/s3-aws/s3.go [hunk 18]
```diff
@@ -1300,12 +1355,12 @@ func (w *writer) Write(p []byte) (int, error) {
...
+		completedUploadedParts := make(completedParts, len(w.parts))
+		for i, part := range w.parts {
+			completedUploadedParts[i] = &s3.CompletedPart{
 				ETag:       part.ETag,
 				PartNumber: part.PartNumber,
-			})
+			}
 		}
 
 		sort.Sort(completedUploadedParts)
```
Comment (thaJeztah) [inline registry/storage/driver/s3-aws/s3.go:1366]: Unrelated: wondering if `sort.Stable` (or `slices.SortStableFunc` as recommended on that function) would bring performance benefits.

File: registry/storage/driver/s3-aws/s3.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply registry/storage/driver/s3-aws/s3.go:1268]: I'm not strongly attached to either, but as this was a straight move, without changes, it could make sense to do that move in a separate commit (to make it clearer what parts of this patch are related to the optimisation, and what parts are to "assist" those changes).

File: registry/storage/driver/s3-aws/s3.go [hunk 19]
```diff
@@ -1499,47 +1575,56 @@ func (w *writer) Commit() error {
...
 		return nil
 	}
-	if w.driver.MultipartCombineSmallPart && len(w.pendingPart) < int(w.driver.ChunkSize) {
-		// closing with a small pending part
-		// combine ready and pending to avoid writing a small part
-		w.readyPart = append(w.readyPart, w.pendingPart...)
-		w.pendingPart = nil
+
+	buf := bytes.NewBuffer(w.ready.data)
+	if w.driver.MultipartCombineSmallPart && (w.pending.Len() > 0 && w.pending.Len() < int(w.driver.ChunkSize)) {
```
Comment (thaJeztah) [inline registry/storage/driver/s3-aws/s3.go:1598]: nit: looks like the extra braces are not needed here

```go
if w.driver.MultipartCombineSmallPart && w.pending.Len() > 0 && w.pending.Len() < int(w.driver.ChunkSize) {
```

File: registry/storage/driver/s3-aws/s3.go [hunk 20]
```diff
@@ -1380,51 +1444,60 @@ func (w *writer) Write(p []byte) (int, error) {
...
-					w.size += int64(n)
-					return n, err
-				}
-			} else {
-				w.pendingPart = append(w.pendingPart, p...)
-				n += len(p)
-				p = nil
+		// try filling up the pending parts buffer
+		offset, err = w.pending.ReadFrom(reader)
+		n += int(offset)
```
Comment (thaJeztah) [inline registry/storage/driver/s3-aws/s3.go:1468]: As we're calling this in a `for`; wondering if it would be beneficial to change `n` to an `int64`, and only cast to an `int` when returning

File: registry/storage/driver/s3-aws/s3.go [hunk 21]
```diff
@@ -1380,51 +1444,60 @@ func (w *writer) Write(p []byte) (int, error) {
...
+		if err != nil {
+			return n, err
```
Comment (thaJeztah) [inline registry/storage/driver/s3-aws/s3.go:1470]: I'd have to dig deeper into the whole flow, but is it intentional to return both `n` and an `err` here? (The code higher up at line 1418 does a `return 0, err` on errors.

File: registry/storage/driver/s3-aws/s3.go [hunk 22]
```diff
@@ -1380,51 +1444,60 @@ func (w *writer) Write(p []byte) (int, error) {
...
-			} else {
-				w.readyPart = append(w.readyPart, p...)
-				n += len(p)
-				p = nil
-			}
+	defer func() { w.size += int64(n) }()
+
+	reader := bytes.NewReader(p)
+
+	for reader.Len() > 0 {
```
Comment (thaJeztah) [inline registry/storage/driver/s3-aws/s3.go:1451]: I guess `reader := bytes.NewReader(p)` could be inlined here, but that's _really_  nitpicking.

**Review by thaJeztah — COMMENTED** [review]
Gave this a first glance, and looks great; I left some questions / comments (none really blocking, afaics)

File: registry/storage/driver/s3-aws/s3.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1268]: This is a fair point @thaJeztah I will put it back so it doesn't pop up in the diff, but as I said my personal preference is to have type implementation grouped together rather than interleaving them with other types 😄

File: registry/storage/driver/s3-aws/s3.go (hunk 20 above)
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1468]: I considered this, indeed. I stuck with this at the time. There are a few returns here so I struggle to see the win 🤔 I'm willing to be persuaded though

File: registry/storage/driver/s3-aws/s3.go (hunk 21 above)
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1470]: I think this is an interesting shout. `writer` is supposed to implement https://pkg.go.dev/io#Writer 

> It returns the number of bytes written from p (0 <= n <= len(p)) and any error encountered that caused the write to stop early. Write must return a non-nil error if it returns n < len(p)

Now, the code on line 1418 does NOT read from `p` which is the `byte` slice passed into `Write`. I think it's fine to return `0, err` there and `n, err` here. WDYT @corhere

File: registry/storage/driver/s3-aws/s3.go (hunk 18 above)
Comment (corhere) [reply registry/storage/driver/s3-aws/s3.go:1366]: No, `sort.Stable` is slower than `sort.Sort`.
> Sort sorts data in ascending order as determined by the Less method. It makes one call to data.Len to determine n and **O(n*log(n)) calls to data.Less and data.Swap.** The sort is not guaranteed to be stable.

> [Stable] makes one call to data.Len to determine n, <strong>O(n\*log(n)) calls to data.Less and O(n*log(n)*log(n)) calls to data.Swap.</strong>

Additionally, the two sort algorithms will produce identical results as every `PartNumber` is unique in the slice.

File: registry/storage/driver/s3-aws/s3.go [hunk 23]
```diff
@@ -1251,21 +1261,70 @@ func (d *driver) getStorageClass() *string {
...
+}
+
+// ReadFrom reads as much data as it can fit in from r without growing its size.
+// It returns the number of bytes successfully read from r or error.
+func (b *buffer) ReadFrom(r io.Reader) (offset int64, err error) {
+	for len(b.data) < cap(b.data) && err == nil {
+		var n int
+		n, err = r.Read(b.data[len(b.data):cap(b.data)])
+		offset += int64(n)
+		b.data = b.data[:len(b.data)+n]
```
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1282]: Yeah, I'm not sure it's worth it.  Maybe? 🤷‍♂️ `len` and `cap` just return an unexported field from the underlying struct so I'd expect negligible gain here 🤔

File: registry/storage/driver/s3-aws/s3.go (hunk 22 above)
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1451]: Yeah, we could probably be more clever in this code but let's keep it as unless you insist 😄

File: registry/storage/driver/s3-aws/s3.go (hunk 18 above)
Comment (thaJeztah) [reply registry/storage/driver/s3-aws/s3.go:1366]: Looks like I had my wires crossed there, and somehow misremembered  that `Stable` did less (so faster) than `Sort`; thanks!

File: registry/storage/driver/s3-aws/s3.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (corhere) [reply registry/storage/driver/s3-aws/s3.go:1282]: No; `len(b.data)` and `cap(b.data)` [are just accessing fields on the slice header](https://pkg.go.dev/reflect@go1.21.1#SliceHeader) which costs the same as any other struct field read. [Compiler explorer reveals](https://go.godbolt.org/z/nq8nrnE6E) that the compiler is able to elide bounds-checks for the expression `b.data[len(b.data):cap(b.data)]` which it cannot do when those values are copied to local variables. Copying the slice itself to a local variable sounds promising as a way to avoid dereferencing pointers inside the loop, but as compiler explorer shows, the generated AMD64 code immediately spills the slice header fields to the stack.

tl;dr introducing intermediate variables will actually deoptimize the function. Compilers are pretty smart these days.

File: registry/storage/driver/s3-aws/s3.go [hunk 24]
```diff
@@ -1251,21 +1261,70 @@ func (d *driver) getStorageClass() *string {
...
+	}
+	// NOTE(milosgajdos): io.ReaderFrom "swallows" io.EOF
+	// See: https://pkg.go.dev/io#ReaderFrom
+	if err == io.EOF {
+		err = nil
+	}
+	return offset, err
```
[the remark this answers is not in the document]
Comment (corhere) [reply registry/storage/driver/s3-aws/s3.go:1289]: [Returning 0 on error would violate the `io.ReaderFrom` interface contract.](https://pkg.go.dev/io#ReaderFrom)

File: registry/storage/driver/s3-aws/s3.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/s3-aws/s3.go:1282]: yes, sorry by "unexported fields" I meant "fields" and yeah, header is the right terminology. Thanks

File: registry/storage/driver/s3-aws/s3.go (hunk 21 above)
Comment (corhere) [reply registry/storage/driver/s3-aws/s3.go:1470]: [Returning the number of bytes written **from `p`** is a requirement of the `io.Writer` interface contract.](https://pkg.go.dev/io#Writer) Zero bytes have been written from `p` if control flow reaches line 1418.

**Comment by milosgajdos** [comment]
PTAL @thaJeztah. All the important points have been addressed.

File: registry/storage/driver/s3-aws/s3.go (hunk 20 above)
Comment (thaJeztah) [reply registry/storage/driver/s3-aws/s3.go:1468]: Yeah, I was curious if benchmarking would show a difference; I just did a quick check, and there's no consistent difference, so not worth changing.

File: registry/storage/driver/s3-aws/s3.go (hunk 21 above)
Comment (thaJeztah) [reply registry/storage/driver/s3-aws/s3.go:1470]: Thanks! That's the part I wanted to dig deeper into, because I suspected that information had to be preserved (even in the error case). I guess the interface is just slightly confusing to require both error and value, but it makes sense in context.

**Review by thaJeztah — APPROVED** [review]
LGTM
